Fixed #38346: Eigen decomposition of Symmetric Matrix containing NaNs now throws exception

kc611 opened the pull request:
Fixes JuliaLang/LinearAlgebra.jl#780.

As suggested by @KlausC in the issue 

> There seems missing a call to `ckfinite` or analogous in `syevr!`

adding a call to 'ckfinite' in 'syevr' is a nice fix for the problem. Since Eigen decomposition for both symmetric and hermitian  matrix is making call to 'syevr' function it fixes the issue for both. Diagonal matrices already had such a fail-safe for NaN values 

https://github.com/JuliaLang/julia/blob/af3cd59e13c93fcb3e3788a79b0168127473d18a/stdlib/LinearAlgebra/src/diagonal.jl#L674-L676

kc611 committed:
Fixed 38346:Eigen Decomposition with NaNs runs forever

kc611 committed:
Merge branch 'master' of https://github.com/JuliaLang/julia

dkarrasch commented:
@kc611 Thank you and welcome! Can you please add a test (à la `@test_throws ...`) to make sure the function throws when it should throw?

kc611 committed:
Added tests

kc611 commented:
It seems that there was already a test present for Eigen of matrices with NaN values. I extended it's functionality to Symmetric, Hermitian and Diagonal(just in case) ones.

KlausC commented on stdlib/LinearAlgebra/src/lapack.jl:
```diff
@@ -5030,6 +5030,7 @@ for (syev, syevr, sygvd, elty) in
                         vl::AbstractFloat, vu::AbstractFloat, il::Integer, iu::Integer, abstol::AbstractFloat)
             chkstride1(A)
             n = checksquare(A)
+            chkfinite(A)
```
I am not sure, that `chfinite` should be called unmodified. As it checks all elements of `A`, that could erroneously throw exceptions, if there is a non-finite element in the unused part of `A`.
Only one triagle of `A` should be checked, dependant on `uplo`.

dkarrasch replied:
Good point.

kc611 replied:
Looks like this will require adding another function similar to chkfinite() something like :

`function chkuplofinite(A::AbstractMatrix, uplo::AbstractChar)`

which will then check upper or lower part of matrix for non-finite element according to `uplo`
Should I go ahead and implement this?

kc611 replied:
@dkarrasch @KlausC I've added the functionality as requested.

JeffBezanson reviewed (changes requested):
Has comments that need to be addressed.

kc611 committed:
Added chkuplofinite : checks the upper or lower part for nonfinite values appropriately

KlausC commented on stdlib/LinearAlgebra/src/lapack.jl:
```diff
@@ -95,6 +95,23 @@ function chkfinite(A::AbstractMatrix)
+function chkuplofinite(A::AbstractMatrix, uplo::AbstractChar)
+    m, n = size(A)
+    if uplo == 'U'
+        for i in 1:m, j in i:n
```
what do you think about changing the loops to iterate column-wise?
` for j = 1:n, i = 1:j`

KlausC commented on stdlib/LinearAlgebra/src/lapack.jl:
```diff
@@ -95,6 +95,23 @@ function chkfinite(A::AbstractMatrix)
+            end
+        end
+    else
+        for i in 1:m, j in 1:i
```
column order would be
`for j = 1:n, i = j:m`

dkarrasch commented on stdlib/LinearAlgebra/src/lapack.jl:
```diff
@@ -95,6 +95,23 @@ function chkfinite(A::AbstractMatrix)
     return true
 end
 
+function chkuplofinite(A::AbstractMatrix, uplo::AbstractChar)
```
To play it safe, I think we should also check that we may indeed assume one-based indexing:
```suggestion
function chkuplofinite(A::AbstractMatrix, uplo::AbstractChar)
    require_one_based_indexing(A)
```

dkarrasch replied:
... and once we have that, the loops below should get a `@inbounds` prepended.

kc611 committed:
Changed iterations to columnwise and added one based indexing check

kc611 commented:
Changed the iterations to column-wise (the loops looks a bit clean and consistent that way) and added one based indexing check and `@inbounds` as suggested.

Also I was thinking about adding a test case where the code is not supposed to throw exception. But dropped the idea since I could not find any appropriate way for doing that (#18780). Any suggestions?

KlausC commented:
To verify, that a test does not throw and exception you could do for example:
```
@test eigen(...) isa Eigen
or
@test something....   !== []
or
@test something...  !== nothing  (if you expect any return type != nothing
```

kc611 committed:
Added more tests

KlausC commented on stdlib/LinearAlgebra/test/eigen.jl:
```diff
@@ -101,11 +101,12 @@ end
         @test_throws(ArgumentError, eigen(Symmetric(test_matrix)))
         @test_throws(ArgumentError, eigen(Hermitian(test_matrix)))
-        @test_throws(ArgumentError, eigen(Diagonal(test_matrix)))
+        @test eigen(Symmetric(transpose(test_matrix))) isa Eigen
+        @test eigen(Hermitian(transpose(test_matrix))) isa Eigen
```
better use `Symmetric/Hermitian(test_matrix, :L)` to test also this code branch.

kc611 committed:
Modified tests appropriately

dkarrasch reviewed (approved):
LGTM!

dkarrasch commented:
The usual unrelated test failures. Thank you very much @kc611, and sorry that it took a while. Your contribution is much appreciated!
